openapi3/internalize_refs.go

```diff
@@ -0,0 +1,373 @@
+package openapi3
+
+import (
+	"context"
+	"strconv"
+	"strings"
+)
+
+func schemaNames(s Schemas) []string {
+	out := make([]string, 0, len(s))
+	for i := range s {
+		out = append(out, i)
+	}
+	return out
+}
+
+func parametersMapNames(s ParametersMap) []string {
+	out := make([]string, 0, len(s))
+	for i := range s {
+		out = append(out, i)
+	}
+	return out
+}
+
+// InternalizeRefs removes all references to external files from the spec and moves them
+// to the components section.
+//
+// refNameResolver takes in references to returns a name to store the reference under locally.
+// Some care should be taken to make sure the resolver does not return duplicate names for different
+// references.
+func (spec *T) InternalizeRefs(ctx context.Context, refNameResolver func(ref string) string) *T {
+	isExternalRef := func(ref string) bool {
+		return ref != "" && !strings.HasPrefix(ref, "#/components/")
+	}
+
+	addSchemaToSpec := func(s *SchemaRef) string {
+		if s == nil || !isExternalRef(s.Ref) {
+			return ""
+		}
+
+		name := refNameResolver(s.Ref)
+
+		// I just added a more comprehensive check for this being a schema but people should really
+		// make sure their resolver returns unique names
+		val := 1
+		for {
+			// basic check, may need something more exhaustive...
+			if existing, ok := spec.Components.Schemas[name]; ok {
+				if existing.Value.Description != s.Value.Description || s.Value.Type != existing.Value.Type {
+					name = refNameResolver(s.Ref) + strconv.Itoa(val)
+					val++
+					continue
+				}
+				newRef := "#/components/schemas/" + name
+				if newRef == s.Ref {
+					s.Ref = ""
+				} else {
+					s.Ref = newRef
+				}
+				return s.Ref
+			}
+			break
+		}
+		if spec.Components.Schemas == nil {
+			spec.Components.Schemas = make(Schemas)
+		}
+		spec.Components.Schemas[name] = s.Value.NewRef()
+		s.Ref = "#/components/schemas/" + name
+		return s.Ref
+	}
+
+	addParameterToSpec := func(p *ParameterRef) string {
```

Comment:
Why not define all these closures as functions of `*T`? e.g.
```go
func (doc *T) addParameterToSpec(p *ParameterRef) string { ... }
```
IMO it would be preferable.

Reply:
No problem at all, it did turn into more of a mess than I was anticipating